Remotion/Web/Core/Resources/ResourcePathBuilderBasedStaticResourceCacheKeyProvider.cs

```diff
@@ -0,0 +1,266 @@
+﻿// This file is part of the re-motion Core Framework (www.re-motion.org)
+// Copyright (c) rubicon IT GmbH, www.rubicon.eu
+//
+// The re-motion Core Framework is free software; you can redistribute it
+// and/or modify it under the terms of the GNU Lesser General Public License
+// as published by the Free Software Foundation; either version 2.1 of the
+// License, or (at your option) any later version.
+//
+// re-motion is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with re-motion; if not, see http://www.gnu.org/licenses.
+//
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Security.Cryptography;
+using System.Text;
+using Remotion.Utilities;
+using Remotion.Web.Infrastructure;
+
+namespace Remotion.Web.Resources
+{
+  /// <summary>
+  /// Defines an API for retrieving the cache key for static resources URLs generated by <see cref="ResourcePathBuilder"/>.
+  /// Changes to the folder are watched and will cause the generation of a new cache key.
+  /// </summary>
+  /// <threadsafety static="true" instance="true" />
+  public class ResourcePathBuilderBasedStaticResourceCacheKeyProvider : IStaticResourceCacheKeyProvider, IDisposable
+  {
+    public interface IMappedResourcesPathLocator
+    {
+      string GetMappedResourcesPath ();
+    }
+
+    public class MappedResourcesPathLocator : IMappedResourcesPathLocator
+    {
+      private readonly IHttpContextProvider _httpContextProvider;
+      private readonly ResourceRoot _resourceRoot;
+
+      public MappedResourcesPathLocator (IHttpContextProvider httpContextProvider, ResourceRoot resourceRoot)
+      {
+        ArgumentUtility.CheckNotNull(nameof(httpContextProvider), httpContextProvider);
+        ArgumentUtility.CheckNotNull(nameof(resourceRoot), resourceRoot);
+
+        _httpContextProvider = httpContextProvider;
+        _resourceRoot = resourceRoot;
+      }
+
+      /// <inheritdoc />
+      public string GetMappedResourcesPath ()
+      {
+        var context = _httpContextProvider.GetCurrentHttpContext();
+        return context.Server.MapPath("~/" + _resourceRoot.Value);
+      }
+    }
+
+    public interface IResourceFileDetailsAppender
+    {
+      void AppendResourceFolderDetails (List<(string Name, long Length, long CreationTime, long LastWriteTime)> folderDetails, string resourceFolder, IReadOnlyCollection<ResourceType> resourceTypes);
```

Comment:
I know I had both this version and the alternative of using a List<string> + List<long> in my suggestions. I realize now that I prefer the Two-List approach because you can pack anything into the cachekey. Does the tuple-version have particular benefits that make it preferable?

Thoughts on the two variants?

Reply:
This is not an extension point but instead is only here to enable testing. As such, I don't think that using a list here adds any benefit. 

I assume you would use `List<List<object>>` instead? That would require casting all the long objects. If you wanted to hack the hash you could just append something to the name.

Reply:
no, I literally meant to parameters, one List<string> the second one List<long>. To me, it would also simplify testing and it looks weird having a public API like that.

If it's not a good idea, we can keep it as is.

Reply:
Can you elaborate on what you mean by "second list"? I don't know what a second list would add here...

Reply:
Ah okay now I understand your idea. I think the current version, while a bit more verbose, is better as each entry (name, length, creation date, and modified date) is stored side by side which makes the tests more readable. In regards to boxing, we don't do any boxing here as well since we use `ValueTuple<T>` so both solutions are equivalent in that they don't box.

Reply:
Exactly. The boxing was just why it's two lists instead of one list of object. Done.